File: src/SIR.Client.Web/RulesExplorer.fs
module SIR.Client.Web.RulesExplorer

open Feliz
open SIR.Client
open SIR.Domain
open SIR.Simulation

let private valueText value =
    match value.Value with
    | IntegerValue number -> string number + " " + value.Unit
    | FixedPointValue number -> string (FixedPoint.raw number) + "/" + string FixedPoint.Scale + " " + value.Unit
    | BooleanValue flag -> string flag + " " + value.Unit
    | TextValue content -> content + " " + value.Unit

let private catalogTable (label: string) (headers: string list) (rows: string list list) =
    Html.div [
        prop.className "catalog-table-scroll"
        prop.children [
            Html.table [
                prop.ariaLabel label
                prop.children [
                    Html.caption label
                    Html.thead [ Html.tr [ for header in headers do Html.th header ] ]
                    Html.tbody [ for row in rows do Html.tr [ for value in row do Html.td value ] ]
                ]
            ]
        ]
    ]

let private rulesCatalog () =
    Html.section [
        prop.children [
            Html.h2 "Units, perks, weapons, and equipment"
            Html.p "Tables label canonical, proposed, and prototype laboratory values."
            Html.details [
                prop.isOpen true
                prop.children [
                    Html.summary "Units and body profiles"
                    catalogTable "Unit roles" [ "Unit"; "Faction"; "Status"; "Role" ] [ for unit in RulesCatalog.unitRoles do [ unit.Name; unit.Faction; unit.Status; unit.Role ] ]
                    catalogTable "Prototype body profiles" [ "Body"; "Status"; "HP"; "Front armor"; "Flank armor"; "Rear armor"; "Suppression resistance"; "Regeneration/s" ] [ for body in RulesCatalog.bodyProfiles do [ body.Name; body.Status; body.Health; body.FrontArmor; body.FlankArmor; body.RearArmor; body.SuppressionResistance; body.RegenerationPerSecond ] ]
                ]
            ]
            Html.details [
                Html.summary ("Perks · " + string RulesCatalog.perkProfiles.Length)
                catalogTable "Perk families" [ "Family"; "Perk"; "Tactical change" ] [ for perk in RulesCatalog.perkProfiles do [ perk.Family; perk.Name; perk.TacticalChange ] ]
            ]
            Html.details [
                Html.summary "Weapons and prototype profiles"
                catalogTable "Canonical weapon roles" [ "Weapon"; "Engagement shape"; "Target"; "Tactical role" ] [ for weapon in RulesCatalog.weaponRoles do [ weapon.Name; weapon.EngagementShape; weapon.Target; weapon.TacticalRole ] ]
                catalogTable "Prototype weapon profiles" [ "Weapon"; "Kind"; "Base engage (s)"; "Range slope"; "Exponent"; "Accuracy"; "Dispersion/m"; "Damage"; "Penetration"; "Shots/s"; "Effect density"; "Suppression/s" ] [ for weapon in RulesCatalog.weaponProfiles do [ weapon.Name; weapon.Kind; weapon.BaseEngageSeconds; weapon.RangeSlope; weapon.Exponent; weapon.Accuracy; weapon.DispersionPerMeter; weapon.Damage; weapon.Penetration; weapon.ShotsPerSecond; weapon.EffectDensity; weapon.SuppressionPerSecond ] ]
            ]
            Html.details [
                Html.summary "Armor and equipment"
                catalogTable "Human armor packages" [ "Package"; "Coverage"; "Cost" ] [ for armor in RulesCatalog.armorProfiles do [ armor.Name; armor.Coverage; armor.Cost ] ]
                catalogTable "Equipment catalog" [ "Faction"; "Status"; "Category"; "Items" ] [ for equipment in RulesCatalog.equipmentGroups do [ equipment.Faction; equipment.Status; equipment.Category; equipment.Items ] ]
            ]
            Html.p [ Html.a [ prop.href "gameplay-reference.html"; prop.text "Read definitions, formulas, and design rationale in the Gameplay Reference." ] ]
        ]
    ]

[<ReactComponent>]
let ExecutableRulesPanel () =
    let resolvedAttack, setResolvedAttack = React.useState<SIR.Simulation.SimulationEvent option>(None)
    Html.section [
        prop.ariaLabel "Rules data tables"
        prop.children [
            Html.h2 ("Executable combat rules · " + SIR.Simulation.CombatRules.packageIdentity.PackageVersion)
            Html.button [
                prop.text "Execute canonical player attack"
                prop.onClick (fun _ ->
                    let result = SIR.Simulation.Simulation.runTick SIR.Simulation.Simulation.initialState SIR.Simulation.Simulation.inputs
                    result.Events
                    |> List.tryFind (fun (event: SIR.Simulation.SimulationEvent) ->
                        match event with
                        | SIR.Simulation.SimulationEvent.AttackResolved _ -> true
                        | _ -> false)
                    |> setResolvedAttack)
            ]
            Html.section [
                prop.ariaLabel "Authoritative attack event"
                prop.children [
                    match resolvedAttack with
                    | Some(SIR.Simulation.SimulationEvent.AttackResolved(_, _, damage, remainingHealth, explanation)) ->
                        Html.h3 ("AttackResolved event " + explanation.EventId)
                        Html.p ("Damage " + string damage + " · remaining health " + string remainingHealth + " · outcome " + valueText explanation.Outcome)
                        let governingId = RuleId.value explanation.RuleId
                        Html.a [ prop.href ("#rule-" + governingId); prop.text ("Open governing rule " + governingId) ]
                        Html.h4 "Decisive applications"
                        Html.ul [
                            for application in explanation.Children do
                                Html.li (RuleId.value application.RuleId + " · " + (application.Operands |> List.map (fun (name, value) -> name + "=" + valueText value) |> String.concat "; ") + " → " + valueText application.Outcome)
                        ]
                    | _ -> Html.p "No authoritative attack has been emitted yet."
                ]
            ]
            for rule in SIR.Simulation.CombatRules.registry do
                let metadata = rule.Metadata
                Html.details [
                    prop.id ("rule-" + RuleId.value metadata.Id)
                    if RuleId.value metadata.Id = "COMBAT-ATTACK-RESOLUTION-001" then prop.isOpen true
                    prop.children [
                        Html.summary (RuleId.value metadata.Id + " · " + metadata.Title)
                        match rule.Semantics with
                        | FormulaSemantics(_, _, expression) -> Html.p (Rules.formulaNotation expression)
                        | FactSemantics value -> Html.p ("Fact value · " + valueText value)
                        | AlgorithmSemantics contract -> Html.p ("Algorithm · " + contract.ImplementationSymbol + " · " + contract.Fingerprint + " · explains " + String.concat ", " contract.ExplanationFields)
                        | TransitionSemantics contract -> Html.p ("Transition · " + contract.Phase + " · events " + String.concat ", " contract.Events)
                        | _ -> ()
                        Html.p (metadata.Rationale + " · dependencies: " + (metadata.Dependencies |> List.map RuleId.value |> String.concat ", ") + " · evidence: " + String.concat "; " metadata.Evidence)
                        Html.p ("examples: " + String.concat "; " metadata.Examples + " · properties: " + String.concat "; " metadata.Properties)
                        Html.a [ prop.href "../../tests/fixtures/rules-corpus/v2/coverage.json"; prop.text "Coverage graph" ]
                        match metadata.RuleSource with
                        | Some source -> Html.a [ prop.href ("https://github.com/EHotwagner/S.I.R./blob/" + source.Commit + "/" + source.RepositoryPath); prop.text ("Pinned F# source · " + source.Symbol) ]
                        | None -> Html.span "No executable source"
                    ]
                ]
        ]
    ]

[<ReactComponent>]
let DeferredDataPanel (simulator: SimulatorHandoff option, selectedUnit: int32 option) =
    let selected =
        simulator
        |> Option.bind (fun state ->
            selectedUnit
            |> Option.bind (fun id -> Map.tryFind id state.RuntimeMap.Units)
            |> Option.map (fun unit -> state, unit))

    Html.div [
        prop.children [
            Html.section [
                prop.ariaLabel "Selected unit spatial diagnostics"
                prop.children [
                    Html.h2 "Spatial diagnostics"
                    match selected with
                    | None -> Html.p "Select a simulator unit to inspect authoritative spatial queries."
                    | Some(state, unit) ->
                        let cell col row: FS.GG.Game.Core.Cell = { Col = col; Row = row }
                        let identity =
                            SpatialAuthorityIdentity.create state.Revision.Digest "sir-spatial-v1" state.Revision.Number "player-disclosed" state.Revision.Number
                            |> Result.defaultWith failwith
                        let world =
                            { Identity = identity
                              Minimum = cell 0 0
                              Maximum = cell (state.RuntimeMap.Width - 1) (state.RuntimeMap.Height - 1)
                              Terrain =
                                state.RuntimeMap.Terrain
                                |> Map.toList
                                |> List.map (fun ((column, row), terrain) ->
                                    cell column row,
                                    match terrain with
                                    | MapTerrain.Open | MapTerrain.Objective -> SpatialTerrain.Open
                                    | MapTerrain.Rough -> SpatialTerrain.Rough
                                    | MapTerrain.Blocked -> SpatialTerrain.Blocked)
                                |> Map.ofList
                              Boundaries = []
                              Occupancy = Map.empty
                              DisclosedRevisionTokens = Set.empty }
                        let origin = cell unit.Column unit.Row
                        let target = cell (min (state.RuntimeMap.Width - 1) (unit.Column + 4)) unit.Row
                        let request =
                            { QueryId = "selected-unit-diagnostics"
                              QueryKind = SpatialQueryKind.ExactLineOfSight
                              Origin = origin
                              Target = target
                              Footprint = [ for row in 0 .. unit.Size - 1 do for column in 0 .. unit.Size - 1 do yield cell column row ]
                              Profile = { ProfileId = "selected-unit-sensor-v1"; Modality = SpatialModality.Vision; Stance = "standing"; HeightBand = 1; Facing = unit.AttentionDirection }
                              Bounds = SpatialQuery.defaultBounds }
                        let result, _ = SpatialQuery.evaluate world request
                        Html.dl [
                            Html.dt "Query"
                            Html.dd (string result.Explanation.QueryKind)
                            Html.dt "Outcome"
                            Html.dd (string result.Outcome)
                            Html.dt "Footprint samples"
                            Html.dd (string result.Explanation.FootprintSamples.Length)
                            Html.dt "Crossed cells / edges"
                            Html.dd (string result.Explanation.CrossedCells.Length + " / " + string result.Explanation.CrossedEdges.Length)
                            Html.dt "Cover contributors"
                            Html.dd (string result.Explanation.CoverContributors.Length)
                            Html.dt "Exposure directions"
                            Html.dd (result.Explanation.ExposureDirections |> List.map string |> String.concat ", ")
                            Html.dt "Revision / knowledge policy"
                            Html.dd (string result.Explanation.SpatialRevision + " / " + result.Explanation.KnowledgeIdentity)
                            Html.dt "Package / profile"
                            Html.dd (SpatialQuery.packageIdentity + " / " + SpatialQuery.compatibilityProfile)
                            Html.dt "Pinned authority"
                            Html.dd result.Explanation.SourceSymbol
                        ]
                ]
            ]
            ExecutableRulesPanel ()
            rulesCatalog ()
        ]
    ]

File: src/SIR.Client.Web/.fable-rules/RulesExplorer.js

import { int32ToString } from "./fable_modules/fable-library-js.5.13.0/Util.js";
import { FixedPointModule_raw } from "./SIR.Domain/FixedPoint.js";
import { toString } from "./fable_modules/fable-library-js.5.13.0/Types.js";
import { createElement, useState } from "react";
import React from "react";
import { HtmlHelper_createElement } from "./fable_modules/Feliz.3.3.3/Html.fs.js";
import { empty, collect, map, singleton, append, delay, toList } from "./fable_modules/fable-library-js.5.13.0/Seq.js";
import { CombatRules_registry, CombatRules_packageIdentity } from "./SIR.Simulation/CombatRules.js";
import { defaultOf } from "./fable_modules/fable-library-js.5.13.0/Util.js";
import { map as map_1, ofArray, tryFind, singleton as singleton_1 } from "./fable_modules/fable-library-js.5.13.0/List.js";
import { Simulation_inputs, Simulation_initialState, Simulation_runTick } from "./SIR.Simulation/Simulation.js";
import { RuleIdModule_value } from "./SIR.Domain/RuleTypes.js";
import { join } from "./fable_modules/fable-library-js.5.13.0/String.js";
import { formulaNotation } from "./SIR.Domain/Rules.js";

function valueText(value) {
    const matchValue = value.Value;
    switch (matchValue.tag) {
        case 1:
            return (((int32ToString(FixedPointModule_raw(matchValue.fields[0])) + "/") + int32ToString(10000)) + " ") + value.Unit;
        case 2:
            return (toString(matchValue.fields[0]) + " ") + value.Unit;
        case 3:
            return (matchValue.fields[0] + " ") + value.Unit;
        default:
            return (int32ToString(matchValue.fields[0]) + " ") + value.Unit;
    }
}

export function ExecutableRulesPanel() {
    const patternInput = useState(undefined);
    const resolvedAttack = patternInput[0];
    return HtmlHelper_createElement("div", singleton_1(["children", toList(delay(() => {
        let value;
        return append(singleton((value = ("Executable combat rules · " + CombatRules_packageIdentity.PackageVersion), createElement("h2", defaultOf(), value))), delay(() => append(singleton(HtmlHelper_createElement("button", ofArray([["children", singleton_1("Execute canonical player attack")], ["onClick", (_arg) => {
            patternInput[1](tryFind((event) => {
                if (event.tag === 3) {
                    return true;
                }
                else {
                    return false;
                }
            }, Simulation_runTick(Simulation_initialState, Simulation_inputs).Events));
        }]]))), delay(() => append(singleton(HtmlHelper_createElement("section", ofArray([["aria-label", "Authoritative attack event"], ["children", toList(delay(() => {
            let value_6;
            let matchResult, damage, explanation, remainingHealth;
            if (resolvedAttack != null) {
                if (resolvedAttack.tag === 3) {
                    matchResult = 0;
                    damage = resolvedAttack.fields[2];
                    explanation = resolvedAttack.fields[4];
                    remainingHealth = resolvedAttack.fields[3];
                }
                else {
                    matchResult = 1;
                }
            }
            else {
                matchResult = 1;
            }
            switch (matchResult) {
                case 0:
                    return append(singleton((value_6 = ("AttackResolved event " + explanation.EventId), createElement("h3", defaultOf(), value_6))), delay(() => {
                        let value_7;
                        return append(singleton((value_7 = ((((("Damage " + int32ToString(damage)) + " · remaining health ") + int32ToString(remainingHealth)) + " · outcome ") + valueText(explanation.Outcome)), createElement("p", defaultOf(), value_7))), delay(() => {
                            const governingId = RuleIdModule_value(explanation.RuleId);
                            return append(singleton(HtmlHelper_createElement("a", ofArray([["href", "#rule-" + governingId], ["children", singleton_1("Open governing rule " + governingId)]]))), delay(() => append(singleton(createElement("h4", defaultOf(), "Decisive applications")), delay(() => {
                                let children_10;
                                return singleton((children_10 = toList(delay(() => map((application) => {
                                    const value_14 = (((RuleIdModule_value(application.RuleId) + " · ") + join("; ", map_1((tupledArg) => ((tupledArg[0] + "=") + valueText(tupledArg[1])), application.Operands))) + " → ") + valueText(application.Outcome);
                                    return createElement("li", defaultOf(), value_14);
                                }, explanation.Children))), createElement("ul", defaultOf(), ...children_10)));
                            }))));
                        }));
                    }));
                default:
                    return singleton(createElement("p", defaultOf(), "No authoritative attack has been emitted yet."));
            }
        }))]]))), delay(() => collect((rule) => {
            const metadata = rule.Metadata;
            return singleton(HtmlHelper_createElement("details", toList(delay(() => append(singleton(["id", "rule-" + RuleIdModule_value(metadata.Id)]), delay(() => append((RuleIdModule_value(metadata.Id) === "COMBAT-ATTACK-RESOLUTION-001") ? singleton(["open", true]) : empty(), delay(() => singleton(["children", toList(delay(() => {
                let value_21;
                return append(singleton((value_21 = ((RuleIdModule_value(metadata.Id) + " · ") + metadata.Title), createElement("summary", defaultOf(), value_21))), delay(() => {
                    let matchValue, value_22, value_24, contract, value_25, contract_1, value_26;
                    return append((matchValue = rule.Semantics, (matchValue.tag === 2) ? singleton((value_22 = formulaNotation(matchValue.fields[2]), createElement("p", defaultOf(), value_22))) : ((matchValue.tag === 0) ? singleton((value_24 = ("Fact value · " + valueText(matchValue.fields[0])), createElement("p", defaultOf(), value_24))) : ((matchValue.tag === 4) ? ((contract = matchValue.fields[0], singleton((value_25 = ((((("Algorithm · " + contract.ImplementationSymbol) + " · ") + contract.Fingerprint) + " · explains ") + join(", ", contract.ExplanationFields)), createElement("p", defaultOf(), value_25))))) : ((matchValue.tag === 3) ? ((contract_1 = matchValue.fields[0], singleton((value_26 = ((("Transition · " + contract_1.Phase) + " · events ") + join(", ", contract_1.Events)), createElement("p", defaultOf(), value_26))))) : (empty()))))), delay(() => {
                        let value_27;
                        return append(singleton((value_27 = ((((metadata.Rationale + " · dependencies: ") + join(", ", map_1(RuleIdModule_value, metadata.Dependencies))) + " · evidence: ") + join("; ", metadata.Evidence)), createElement("p", defaultOf(), value_27))), delay(() => {
                            let value_28;
                            return append(singleton((value_28 = ((("examples: " + join("; ", metadata.Examples)) + " · properties: ") + join("; ", metadata.Properties)), createElement("p", defaultOf(), value_28))), delay(() => append(singleton(HtmlHelper_createElement("a", ofArray([["href", "../../tests/fixtures/rules-corpus/v2/coverage.json"], ["children", singleton_1("Coverage graph")]]))), delay(() => {
                                const matchValue_1 = metadata.RuleSource;
                                if (matchValue_1 == null) {
                                    return singleton(createElement("span", defaultOf(), "No executable source"));
                                }
                                else {
                                    const source = matchValue_1;
                                    return singleton(HtmlHelper_createElement("a", ofArray([["href", (("https://github.com/EHotwagner/S.I.R./blob/" + source.Commit) + "/") + source.RepositoryPath], ["children", singleton_1("Pinned F# source · " + source.Symbol)]])));
                                }
                            }))));
                        }));
                    }));
                }));
            }))])))))))));
        }, CombatRules_registry)))))));
    }))]));
}

